src/core/metrics.cc

```diff
@@ -252,20 +254,20 @@ void impl::remove_registration(const metric_id& id) {
     }
 }
 
-void unregister_metric(const metric_id & id) {
-    get_local_impl()->remove_registration(id);
+void unregister_metric(const metric_id & id, int handle) {
+    get_local_impl(handle)->remove_registration(id);
```

Comment:
I think remove registration needs the handle. I wonder if it's possible to bury the handle a little deeper so that it doesn't need to be repeated.

Reply:
The dispatch is required for correctness here. `unregister_metric` is called by the destructor of `metric_groups`. If we remove the handle (and implicitly use the default impl), the destruction of a `metric_groups` that specified a non-zero handle can lead to metrics being deleted from the implementation with the `default_handle`.

For example:
```
metric_groups default;
default.add_group("kafka", sm::make_gauge("foo"), [] { return 0; } sm::description("foo"), {});

{
metric_groups special(100);
special.add_group("kafka", sm::make_gauge("foo"), [] { return 0; } sm::description("foo"), {});

// When special goes out of scope it would delete the metric from 'default' as they have the 
// same group name, metric type and labels.
}
```

Reply:
No I mean this:
```
void impl::remove_registration(const metric_id& id) {
    auto i = get_value_map().find(id.full_name());
```
Should take the handle to lookup the correct value_map.

Reply:
Ok. I looked at this again. `remove_registration` is a member function of `impl`, which means that `get_value_map` in `remvove_registration` is also a member function (defined [here](https://github.com/redpanda-data/seastar/blob/020c9ba4e709f816110d7ddf819fd97dd6833c69/include/seastar/core/metrics_api.hh#L331)). The dispatch was already done at that point so i think it's fine.

Reply:
I can't do name lookup in my head.